### ISSUE TITLE
Describe what happens when submitFrame is not called every frame.

### ISSUE THREAD

**Issue by toji** [w3c/webvr#69]
Inspired by #58, there should be some documentation of what the expected results are when you don't call `submitFrame` at the headset's native framerate.

For the short term it's probably enough to note that the last submitted frame will be re-projected. Looking forward, though, to a time when WebGL is not the soul VR source some interesting use cases emerge: A `VRLayer` that described placing a video element at a certain point in space, for example, could update automatically without any Javascript interaction. (This type of layer could also take advantage of DRM content that can't be rendered normally in WebGL.) At that point it feels redundant to repeatedly call submitFrame, and waiting on it may introduce jank where none is necessary. The same could go for things like a static skybox or the "loading" spinner described in #58. Since there's no script interaction necessary taking JS out of the render loop is beneficial. 

I'm not sure what the right language is to describe it, but in that context it seems like submitFrame is primarily a mechanism to inform the browsers VR compositor that programatically generated contents (WebGL, 2D canvas?) have been updated. (And to continue allowing for greater control over content mirroring for those sources.)

Just some random thoughts. None of this is v1 material, aside from the aforementioned note about re-projection.

**Comment by dmarcos** [comment]
This is my attempt at breaking down the different scenarios where requestFrame might not be called as expected:

1. Content does not call requestPresent at expected framerate (90fps) but just a few frames are dropped. Last-frame is re-projected. 

2. If requestPresent calls drop below a certain threshold where re-projection might be jarring. A spinner or an hourglass is displayed until requestPresent is invoked above the threshold. If low framerate persists for a long time an `slow site` indication might be shown in the headset.

3. If requestPresent stops being called. There are several cases:

 a) There was an exception / error on the current site. The user agent should somehow notify the user in the headset.

 b) The user agent has transitioned to a different VR site. A loading spinner should be rendered to notify the situation. When the new site loads and starts calling requestPresent there should be a consistent way to convey that the user agent has transitioned to a new site before handing off the display to content (displaying the new URL might be appropriate).

 c) The user agent transitions to a site that does not have VR content. We might need a mechanism for sites to statically declare themselves as VR (metatag?). If that hint is not present the user agent displays a message in the headset indicating that the current URL does not have VR content available.

 d) The user agent transitions to a site that returns an HTTP error. An error notification is displayed in the headset.

feedback @kearwood?

**Comment by dmarcos** [comment]
@toji Does the break down look reasonable to you?

**Comment by kearwood** [comment]
The tiered system presented by @dmarcos is what I have in mind as well.  Essentially, we could tolerate a few dropped frames but the UA needs to intervene before lower level protection, such as the HTC Vive default environment rendering, kicks in.  When the UA detects that a site is continuously dropping frames to a degree that may make a user feel uncomfortable, a transparent indication (i.e. a HUD with a notification icon) could be displayed without obscuring the users HMD.

To avoid dropping to the default HTC Vive environment which may confuse users or when bad performance persists long enough to make users feel ill, the UA could then present opaque UX that protects the user from seeing the content and gives them an option sensitive to the context:

- If the user had navigated from another page, the option to return to the prior page.
- To continue waiting for the site to improve performance
- If at least one frame had been submitted, to dismiss the opaque UX and see the content that is performing poorly.
- To exit VR mode.  The UA may trigger an onvrdisplaydeactivate event and enter a "VR Shell" view of the 2d page.

**Comment by kearwood** [comment]
In the case of future VRLayer's that can render themselves without submitFrame calls there could be some special cases:

- If all VRLayer's in the presentation require submitFrame, first present the transparent HUD warning when performance is below threshold, then present the opaque UX with user options if the site does not recover within a period of time.

- If none of the VRLayer's in the presentation require submitFrame, then performance failures may still occur; however, as the UA is in full control it may take measures such as lowering resolution or disabling antialiasing features until resorting to the same HUD warning and eventual opaque UX.

- If the presentation contains a mix of VRLayer's that require submitFrame and those that do not, then the question is whether the VRLayer's rendered by the UA should continue to be presented even if submitFrame calls fail to be called frequently enough for a comfortable experience.  I would propose that this be a choice made by the content authors, with an attribute in the WebGL based layers indicating that they can be suppressed when performance drops or Javascript stops submitting all together.  Rather than just dropping these WebGL layers and treating them as transparent, another VRLayer could be identified by content as a substitute in the event of failure.  If we were to have a simple VRLayer that fills the environment with a solid color or a cube-map, perhaps this would be suitable for this use.

**Comment by toji** [comment]
I agree with all of the above, really. Seems like a very sensible break down of the expected scenarios and handling options.

**Comment by toji** [comment]
Closing since this is planned to be addressed in 2.0. If anyone has issues with the 2.0 proposal, please file a new issue. For details see: https://github.com/w3c/webvr/blob/master/explainer.md#main-render-loop
